server/aesdsocket.c

```diff
@@ -56,128 +82,163 @@ int create_server_socket()
-        exit(EXIT_FAILURE);
+/* ---- Graceful shutdown ---- */
+void shutdown_all()
+{
+    if (sockfd_global != -1) {
+        close(sockfd_global);
+        sockfd_global = -1;
     }
 
-    chdir("/");
-
-    // redirect stdio to /dev/null
-    freopen("/dev/null", "r", stdin);
-    freopen("/dev/null", "w", stdout);
-    freopen("/dev/null", "w", stderr);
+    while (!SLIST_EMPTY(&thread_head)) {
+        struct client_thread *node = SLIST_FIRST(&thread_head);
+        SLIST_REMOVE_HEAD(&thread_head, entries);
+        pthread_join(node->tid, NULL);
+        free(node);
+    }
 }
 
+/* ---- Main ---- */
 int main(int argc, char *argv[])
 {
-    int daemon_mode = 0;
-    if (argc == 2 && strcmp(argv[1], "-d") == 0) {
-        daemon_mode = 1;
-    }
+    int daemon_mode = (argc == 2 && strcmp(argv[1], "-d") == 0);
 
     openlog("aesdsocket", LOG_PID, LOG_USER);
-
     signal(SIGINT, handle_signal);
     signal(SIGTERM, handle_signal);
 
-    // clear file at program start
     FILE *fp = fopen(DATAFILE, "w");
     if (fp) fclose(fp);
 
-    int sockfd = create_server_socket();
-    if (sockfd < 0) return -1;
-    sockfd_global = sockfd;
+    sockfd_global = create_server_socket();
+    if (sockfd_global < 0) return EXIT_FAILURE;
 
     if (daemon_mode) {
-        daemonize();
+        if (fork() > 0) exit(0);
+        setsid();
+        if (fork() > 0) exit(0);
+        chdir("/");
+        close(STDIN_FILENO);
+        close(STDOUT_FILENO);
+        close(STDERR_FILENO);
+    }
+
+    if (pthread_create(&timestamp_thread, NULL, timestamp_func, NULL) != 0) {
+        syslog(LOG_ERR, "Failed to create timestamp thread");
+        exit(EXIT_FAILURE);
     }
 
     while (!exit_requested) {
         struct sockaddr_in client_addr;
-        socklen_t client_len = sizeof(client_addr);
-
-        int clientfd = accept(sockfd, (struct sockaddr *)&client_addr, &client_len);
+        socklen_t len = sizeof(client_addr);
+        int clientfd = accept(sockfd_global, (struct sockaddr *)&client_addr, &len);
         if (clientfd < 0) {
             if (exit_requested) break;
-            syslog(LOG_ERR, "accept failed: %s", strerror(errno));
             continue;
         }
 
-        handle_client(clientfd, &client_addr);
-        close(clientfd);
+        struct client_thread *node = malloc(sizeof(*node));
+        if (!node) {
+            syslog(LOG_ERR, "malloc failed");
+            close(clientfd);
+            continue;
+        }
+        node->clientfd = clientfd;
+        node->client_addr = client_addr;
+
+        SLIST_INSERT_HEAD(&thread_head, node, entries);
+        if (pthread_create(&node->tid, NULL, client_handler, node) != 0) {
+            syslog(LOG_ERR, "pthread_create failed");
+            SLIST_REMOVE(&thread_head, node, client_thread, entries);
+            free(node);
+            close(clientfd);
+        }
     }
 
-    if (sockfd_global != -1) {
-        close(sockfd_global);
-    }
+    shutdown_all();
```

Comment:
The limitation when only cleaning up here is that all memory associated with completed threads remains until exit.